src/index.ml

```diff
@@ -167,18 +170,151 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
 
   let iter_io ?min ?max f io = iter_io_off ?min ?max (fun _ e -> f e) io
 
-  type window = { buf : bytes; off : int64 }
-
-  let get_entry io ~window off =
-    match window with
-    | None ->
-        let buf = Bytes.create entry_size in
-        let n = IO.read io ~off buf in
-        assert (n = entry_size);
-        decode_entry buf 0
-    | Some w ->
-        let off = Int64.(to_int @@ sub off w.off) in
-        decode_entry w.buf off
+  module Entry = struct
+    type t = entry
+
+    module Key = K
+    module Value = V
+
+    let to_key e = e.key
+
+    let to_value e = e.value
+  end
+
+  module IOArray = struct
+    type buffer = {
+      buf : bytes;
+      low_off : int64;
+      high_off : int64;
+      generation : int64;
+    }
+
+    type t = { io : IO.t; mutable buffer : buffer option }
+
+    let v io =
+      Logs.debug (fun m -> m "NEW IOARRAY");
+      { io; buffer = None }
+
+    let get_entry_from_io io off =
+      let buf = Bytes.create entry_size in
+      let n = IO.read io ~off buf in
+      assert (n = entry_size);
+      decode_entry buf 0
+
+    let ( -- ) = Int64.sub
+
+    let get_entry_from_buffer buf off =
+      let buf_off = Int64.(to_int @@ (off -- buf.low_off)) in
+      assert (buf_off <= Bytes.length buf.buf);
+      decode_entry buf.buf buf_off
+
+    type elt = entry
+
+    let is_in_buffer t off =
+      match t.buffer with
+      | None -> false
+      | Some b ->
+          let gen = IO.get_generation t.io in
+          Int64.equal gen b.generation
+          && Int64.compare off b.low_off >= 0
+          && Int64.compare off b.high_off <= 0
+
+    let get t i =
+      let off = Int64.(mul i entry_sizeL) in
+      match t.buffer with
+      | Some b when is_in_buffer t off -> (
+          try
+            Logs.debug (fun m -> m "Getting entry %Ld from buffer" i);
+            let e = get_entry_from_buffer b off in
+            Logs.debug (fun m -> m "Got entry %Ld from buffer" i);
+            e
+          with _ -> assert false )
+      | _ -> get_entry_from_io t.io off
+
+    let length t = Int64.(div (IO.offset t.io) entry_sizeL)
+
+    let set_buffer t ~low ~high ~range =
+      let buf = Bytes.create range in
+      assert (
+        let n = IO.read t.io ~off:low buf in
```

Comment:
After a debugging session with @CraigFe: 
`low` is an index in the array abstraction here, while we would expect and offset, multiplying by `entry_size` is necessary.

Reply:
Resolved by https://github.com/mirage/index/pull/43/commits/114b1a3c002db426dae61f1830f22a7b1d438bd3. Thanks for your help with debugging this.